src/Storages/System/StorageSystemDNSCache.cpp

```diff
@@ -0,0 +1,61 @@
+#include <Access/ContextAccess.h>
+#include <Interpreters/Context.h>
+#include <Common/DNSResolver.h>
+#include <DataTypes/DataTypeDateTime.h>
+#include <DataTypes/DataTypeEnum.h>
+#include <DataTypes/DataTypeString.h>
+#include <DataTypes/DataTypesNumber.h>
+#include "StorageSystemDNSCache.h"
+
+namespace DB
+{
+
+static DataTypePtr getIPFamilyEnumType()
+{
+    return std::make_shared<DataTypeEnum8>(
+    DataTypeEnum8::Values
+        {
+            {"IPv4",           static_cast<Int8>(Poco::Net::AddressFamily::IPv4)},
+            {"IPv6",           static_cast<Int8>(Poco::Net::AddressFamily::IPv6)},
+            {"UNIX_LOCAL",     static_cast<Int8>(Poco::Net::AddressFamily::UNIX_LOCAL)},
+        });
+}
+
+ColumnsDescription StorageSystemDNSCache::getColumnsDescription()
+{
+    return ColumnsDescription
+        {
+            {"hostname",           std::make_shared<DataTypeString>(), "Hostname."},
+            {"ip_address",         std::make_shared<DataTypeString>(), "IP address."},
+            {"ip_family",          getIPFamilyEnumType(), "IP address family."},
+            {"cached_at",          std::make_shared<DataTypeDateTime>(), "Record cached timestamp."},
+        };
+}
+
+void StorageSystemDNSCache::fillData(MutableColumns & res_columns, ContextPtr, const SelectQueryInfo &) const
+{
+    using HostIPPair = std::pair<std::string, std::string>;
+    std::set<HostIPPair> reported_elements;
+
+    for (const auto & [hostname, entry] : DNSResolver::instance().cacheEntries())
+    {
+        for (const auto &address : entry.addresses)
+        {
+            std::string ip = address.toString();
+
+            // Cache might report the same ip address multiple times. Report only one of them.
```

Comment:
Why is it possible?

Reply:
tbh I don't know, but I don't think we should carry it in this PR